psd-web/app/controllers/users/sessions_controller.rb

```diff
@@ -54,5 +77,13 @@ def add_wrong_credentials_errors
       resource.errors.add(:email, I18n.t(:wrong_email_or_password, scope: "sign_user_in.email"))
       resource.errors.add(:password, nil)
     end
+
+    def user_missing_2fa_mobile_verification?(user)
+      Rails.configuration.two_factor_authentication_enabled && user && !user.mobile_number_verified
+    end
+
+    def set_dummy_devise_resource_for_display_errors
```

Comment:
Naming suggestion (I think "dummy" is a bit colloquial?):

```suggestion
    def set_resource_as_new_user_from_params
```

Reply:
Sure! I didn't think much about the name :sweat_smile: